src/contracts/libraries/GPv2Encoding.sol

```diff
@@ -170,26 +177,29 @@ library GPv2Encoding {
             s := calldataload(add(encodedTrade.offset, 172))
         }
 
-        trade.order.typeHash = ORDER_TYPE_HASH;
         trade.order.sellToken = tokens[sellTokenIndex];
         trade.order.buyToken = tokens[buyTokenIndex];
-        trade.order.kind = OrderKind(flags & 0x1);
-        trade.order.partiallyFillable = (flags >> 1) & 0x01 == 0x01;
+        trade.order.kind = OrderKind(flags & 0x01);
+        trade.order.partiallyFillable = flags & 0x02 != 0;
 
         trade.sellTokenIndex = sellTokenIndex;
         trade.buyTokenIndex = buyTokenIndex;
 
-        // NOTE: In order to avoid memory allocation and copying of the order
-        // parameters, use the memory region reserved by the caller for the
-        // order result to effeciently compute the hash in place. Furthermore,
-        // structs are not packed in Solidity, and there are 10 [`Order`] fields
-        // to hash for a total of `10 * sizeof(uint) = 320` bytes. This digest
-        // is a EIP-712 struct hash for the [`Order`] parameters.
+        // NOTE: Compute the EIP-712 order struct hash in place. The hash is
+        // computed from the order type hash concatenated with the ABI encoded
+        // order fields for a total of `10 * sizeof(uint) = 320` bytes.
+        // Fortunately, since Solidity memory structs **are not** packed, they
+        // are already laid out in memory exactly as is needed to compute the
+        // struct hash, just requiring the order type hash to be temporarily
+        // writen to the memory slot coming right before the order data.
         bytes32 orderDigest;
         // solhint-disable-next-line no-inline-assembly
         assembly {
-            let order := mload(trade)
-            orderDigest := keccak256(order, 320)
+            let dataStart := sub(mload(trade), 32)
+            let temp := mload(dataStart)
+            mstore(dataStart, ORDER_TYPE_HASH)
+            orderDigest := keccak256(dataStart, 320)
+            mstore(dataStart, temp)
```

Comment:
Is this change actually improving gas consumption? We are replacing a `mload`/`mstore` pair with two `mload`/`mstore` pairs. Since the full order information is dropped after decoding anyway, we don't really get any long-term benefit of using one memory slot less in the order.

Reply:
No there is no real benefit in terms of gas. Its more to not have the `typeHash` field in the order struct that gets decoded. The idea is to contain the "hackiness" to the `decodeTrade` function instead of having it "leak" into the `Order` struct definition.